spmd/tensor/parallel/api.py

```diff
@@ -86,3 +87,144 @@ def _shard_self_attn_params(name: str, module: nn.Module) -> None:
+            :class:``nn.Linear`` object to be parallelized.
+        device_mesh (DeviceMesh):
+            :class:``DeviceMesh`` object which contains
+            how we distribute tensor across GPUs.
+        weight_distribute_spec (Sequence[Placement]):
+            the tensor distribute spec of DTensor for weight.
+        bias_distribute_spec (Sequence[Placement]):
+            the tensor distribute spec of DTensor for bias.
+
+    Return:
+        None
+    """
+    weight = nn.Parameter(
+        distribute_tensor(module.weight, device_mesh, weight_distribute_spec)
+    )
+    bias = nn.Parameter(
+        distribute_tensor(module.bias, device_mesh, bias_distribute_spec)
+    )
+    module.register_parameter("weight", weight)
+    module.register_parameter("bias", bias)
+
+
+def _parallelize_mlp(
+    module: nn.Module,
+    device_mesh: DeviceMesh,
+    parallel_style: ParallelStyle = PairwiseParallel(),
+    tp_mesh_dim: int = 0,
+) -> None:
+    """
+    This function assumes the input module is a sequence of nn.Linear
+    and we parallelize the module based on the given parallel style.
+    We don't change the FQN of each sub-module and replace each parameter
+    in place.
+
+    Args:
+        module (nn.Module):
+            :class:``nn.Module`` object to be parallelized.
+        device_mesh (DeviceMesh):
+            :class:``DeviceMesh`` object which contains
+            how we distribute tensor across GPUs.
+        parallel_style (ParallelStyle):
+            :class:``ParallelStyle`` object which contains how
+            we prepare input/output for Tensor Parallelism.
+        tp_mesh_dim (int):
+            the dimension of ``device_mesh`` where we perform
+            Tensor Parallelism on.
+
+    Return:
+        None
+
+    .. warning::
+        We now only support ``PairwiseParallel`` for now.
+    """
+
+    # Define hook functions needed for preparing Input/Output.
+    def _module_forward_pre_hook(*args):  # pyre-ignore[2, 3]
+        return args[0](args[-1][0], *args[1:-2])
+
+    def _module_forward_hook(*args):  # pyre-ignore[2, 3]
+        return args[0](args[-1], *args[1:-3])
+
+    if not isinstance(parallel_style, PairwiseParallel):
+        raise NotImplementedError(
+            "Only support PairwiseParallel for MLP parallelization."
+        )
+
+    if not _has_even_num_linears(module):
+        raise RuntimeError("We only support even number of Linear for MLP.")
+
+    if device_mesh.ndim > 1:
+        device_mesh = _create_1d_device_mesh(device_mesh, tp_mesh_dim)
+
+    for i, m in enumerate(
+        filter(lambda x: isinstance(x, nn.Linear), module.children())
+    ):
+        if i % 2 == 0:
+            # Col-wise Parallelize the linear layer
+            _distribute_linear_module(
+                m,  # pyre-ignore[6]  # type: ignore[arg-type]
+                device_mesh,
+                [Shard(0)],
+                [Shard(0)],
+            )
+            m.register_forward_pre_hook(
```

Comment:
shouldn't we only register pre_hook on the first layer, and register post hook on the last layer? If `_parallel_mlp` works on 8 linear layers, we should only register hook for module 0 and module 7?

Also, I'm wondering if we could directly use `distribute_module` on MLPModule, as that'll only register prehook and posthook once, and maybe it also makes easier for pre/post hook?

Reply:
Hmmm I don't have strong opinion for whether to use `distribute_module` or not. But the thing is that how are we going to write `partition_fn`? Unless we also do this layer by layer.

Reply:
NVM, found a solution.